### DISCUSSION TITLE
Transfer learning problem

### DISCUSSION THREAD

**Discussion by Ranakhan5** [mrdbourke/zero-to-mastery-ml#131] [Q&A]
Why is this code not working? In the video, it is working just fine. but when i try, it is showing error

IMAGE_SHAPE = (224, 224)
BATCH_SIZE = 32
def create_model(model_url, num_classes=10):
  """Takes a TensorFlow Hub URL and creates a Keras Sequential model with it.
  
  Args:
    model_url (str): A TensorFlow Hub feature extraction URL.
    num_classes (int): Number of output neurons in output layer,
      should be equal to number of target classes, default 10.

  Returns:
    An uncompiled Keras Sequential model with model_url as feature
    extractor layer and Dense output layer with num_classes outputs.
  """
  # Download the pretrained model and save it as a Keras layer
  feature_extractor_layer = hub.KerasLayer(model_url,
                                           trainable=False, # freeze the underlying patterns
                                           name='feature_extraction_layer',
                                           input_shape=IMAGE_SHAPE+(3,)) # define the input image shape
  
  # Create our own model
  model = tf.keras.Sequential([
    feature_extractor_layer, # use the feature extraction layer as the base
    layers.Dense(num_classes, activation='softmax', name='output_layer') # create our own output layer      
  ])

  return model
  
  # Create model
resnet_model = create_model(resnet_url, num_classes=train_data_10_percent.num_classes)

# Compile
resnet_model.compile(loss='categorical_crossentropy',
                     optimizer=tf.keras.optimizers.Adam(),
                     metrics=['accuracy'])

**Comment by janeesha-codes** [comment]
# **Building the model**

Before we build a model, there are few things we need to define:
* The input shape (our images shapes in the form of tensors)
* The output shape (image labels in the form of tensors)
* The **URL** of the model we want to use. (Transfer Learning)

```
# Setup the output shape of the model
INPUT_SHAPE = [IMAGE_SIZE, IMAGE_SIZE, 3] #[batch, height, width, color_channels]

# Setup the input shape of the model
OUTPUT_SHAPE = len(unique_breeds)

# Setup model URL from TensorFlow Hub
MODEL_URL = "https://www.kaggle.com/models/google/mobilenet-v2/TensorFlow2/140-224-classification/2"
```
<hr>

# **Create a function to build a Keras model**

* Takes the input shape, output shape and the model we've chosen as parameters (model url).

* Defines the layers in a Keras model in sequential fashion. (step by step approach)

* Compiles the model

* Builds the model

* Returns the model.

All of these steps can be found here: [https://www.tensorflow.org/guide/keras](url)

<hr>

# **tf_keras**

tf_keras is a Keras API packaged separately but fully compatible with TensorFlow.

We will be using `tf_keras` except using `tf.keras` to avoid version miss match errors in TensorFlow Hub.

Some TensorFlow Hub models (like MobileNet, EfficientNet on Kaggle) are built to work specifically with tf_keras.

When to use `tf_keras`?

* When using TensorFlow Hub models
* Pre-trained Kaggle models
* Custom Training Loops

<hr>

```
# Imports tf_keras
import tf_keras
```
```
'''@parameters
input_shape - shape of the input images (e.g. (224, 224, 3))
output_shape - number of output classes (e.g. 120 dog breeds)
model_url - url of the pretrained model from TensorFlow Hub
'''

def create_model(input_shape=INPUT_SHAPE, output_shape=OUTPUT_SHAPE, model_url=MODEL_URL):
  print("Building model with:", model_url)

  # Loads the pretrained DL model from TensorFlow Hub. Wraps it as a Keras layer so it can be used inside a Sequential model.

  '''
  hub.KerasLayer(trainable = False)
   - Freezes the pretrained weights
   - The model acts as a fixed feature extractor
   - Prevents overfitting and reduces training time
  '''

  hubs_layer = hub.KerasLayer(MODEL_URL, trainable=False, input_shape=INPUT_SHAPE)

  '''
  units=OUTPUT_SHAPE
  - One neuron per class. (120 dog breeds)

  ** activation function **
  1. sigmoid (binary classification)
  2. softmax (multi-class classification)
  - Converts outputs into class probabilities
     - Ensures the sum of probabilities equals 1.
  '''

  # Setup the model layers
  model = tf_keras.Sequential([
      hubs_layer, # Layer 1: Pretrained feature extractor (input layer)
      tf_keras.layers.Dense(units=OUTPUT_SHAPE, activation="softmax") # Layer 2 (Output Layer)
  ])

  '''
  ** loss **
  loss - how well the model is guessing (higher the loss the worst the models learning patterns (predictions))

  1. Binary Crossentropy() - (binary classification)

  2. CategoricalCrossentropy()- (multi-class classification)
  Computes the cross-entropy loss between true labels and predicted labels.
  When to use,
  - Used for multi-class classification
     - Assumes labels are one-hot encoded
        Example: [0, 0, 1, 0, 0]

  ** Optimizer **
  Adam() - Adaptive learning rate optimizer
  Efficient and widely used for deep learning tasks

  ** Metrics **
  Tracks how often predictions match true labels during training and validation.
  '''

  # Compile the model
  model.compile(
      loss=tf_keras.losses.CategoricalCrossentropy(),optimizer=tf_keras.optimizers.Adam(),
      metrics=["accuracy"]
    )

  # Build the model
  model.build(INPUT_SHAPE)

  return model
```

```
model_1 = create_model()
model_1.summary()
```
